python-lib/api_calling_utils.py

```diff
@@ -0,0 +1,306 @@
+# -*- coding: utf-8 -*-
+import logging
+import inspect
+import math
+import json
+
+from collections import namedtuple
+from typing import Callable, AnyStr, List, Tuple, NamedTuple, Dict, Union
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+import pandas as pd
+from more_itertools import chunked, flatten
+from tqdm.auto import tqdm as tqdm_auto
+
+from param_enums import ErrorHandlingEnum
+
+# ==============================================================================
+# CONSTANT DEFINITION
+# ==============================================================================
+
+API_COLUMN_NAMES = ["response", "error_message", "error_type", "error_raw"]
+COLUMN_PREFIX = "api"
+
+PARALLEL_WORKERS = 4
+BATCH_SIZE = 10
+
+BATCH_RESULT_KEY = None
+BATCH_ERROR_KEY = None
+BATCH_INDEX_KEY = None
+BATCH_ERROR_MESSAGE_KEY = None
+BATCH_ERROR_TYPE_KEY = None
+
+API_EXCEPTIONS = Exception
+try:
+    from requests.exceptions import RequestException
+    API_EXCEPTIONS = RequestException
+except ImportError:
+    pass
+try:
+    from boto3.exceptions import Boto3Error
+    from botocore.exceptions import BotoCoreError, ClientError
+    API_EXCEPTIONS = (Boto3Error, BotoCoreError, ClientError)
+except ImportError:
+    pass
+try:
+    from google.api_core.exceptions import GoogleAPICallError, RetryError
+    API_EXCEPTIONS = (GoogleAPICallError, RetryError)
+except ImportError:
+    pass
+
+# ==============================================================================
+# FUNCTION DEFINITION
+# ==============================================================================
+
+ApiColumnNameTuple = namedtuple("ApiColumnNameTuple", API_COLUMN_NAMES)
+
+
+def generate_unique(
+    name: AnyStr,
+    existing_names: List,
+    prefix: AnyStr = COLUMN_PREFIX
+) -> AnyStr:
+    """
+    Generate a unique name among existing ones by suffixing a number.
+    Can also add an optional prefix.
+    """
+    if prefix is not None:
+        new_name = prefix + "_" + name
+    else:
+        new_name = name
+    for j in range(1, 1000):
+        if new_name not in existing_names:
+            return new_name
+        new_name = name + "_{}".format(j)
+    raise Exception("Failed to generated a unique name")
```

Comment:
putain le jour ou ça arrive, on ouvre une bouteille de champ.

Reply:
tkt, j'augmenterai a 1001

Reply:
increased to 1001, we are safe now